File: kotlin/src/main/kotlin/phase1/BotLogic.kt
package phase1

import net.folivo.trixnity.core.model.EventId
import net.folivo.trixnity.core.model.UserId
import net.folivo.trixnity.core.model.events.m.ReactionEventContent
import net.folivo.trixnity.core.model.events.m.RelatesTo

data class MirroredReaction(
    val eventId: EventId,
    val key: String,
)

object BotLogic {
    fun mirroredBody(original: String): String = original.uppercase()

    fun shouldHandleSender(sender: UserId, botUserId: UserId): Boolean = sender != botUserId

    fun reactionToMirror(content: ReactionEventContent): MirroredReaction? {
        val relatesTo = content.relatesTo
        if (relatesTo is RelatesTo.Annotation) {
            val key = relatesTo.key ?: return null
            return MirroredReaction(relatesTo.eventId, key)
        }
        return null
    }
}

File: dev/ol/trixnity_poc/bot_logic.clj
(ns ol.trixnity-poc.bot-logic
  (:require
   [clojure.string :as str])
  (:import
   (net.folivo.trixnity.core.model UserId)
   (net.folivo.trixnity.core.model.events.m ReactionEventContent RelatesTo$Annotation)))

(defn mirrored-body [original]
  (str/upper-case original))

(defn should-handle-sender? [^UserId sender ^UserId bot-user-id]
  (not= sender bot-user-id))

(defn reaction-to-mirror [^ReactionEventContent content]
  (let [rel (.getRelatesTo content)]
    (when (instance? RelatesTo$Annotation rel)
      (let [key (.getKey ^RelatesTo$Annotation rel)]
        (when key
          {:event-id (.getEventId ^RelatesTo$Annotation rel)
           :key      key})))))
